### PR TITLE
[SM-1425] Update WASM SDK

### PR DESCRIPTION
**Pull request by coltonhurst**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/SM-1425

## 📔 Objective

Update the WASM SDK to include the SM updates, with the exception of state files.

Sourced from: https://github.com/bitwarden/sdk/pull/1038

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation
  team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: languages/js/example/index.js [hunk 1]
```diff
@@ -11,22 +11,46 @@ async function main() {
     new BitwardenClientWasm(JSON.stringify(settings), LogLevel.Debug),
   );
 
+  await client.loginAccessToken(process.env.ACCESS_TOKEN);
   const organization_id = process.env.ORGANIZATION_ID;
-  await client.accessTokenLogin(process.env.ACCESS_TOKEN);
 
-  const project = await client.projects().create("test", organization_id);
+  const project = await client.projects().create(organization_id, "project-name");
   const projects = await client.projects().list(organization_id);
```
Comment (mzieniukbw) [inline languages/js/example/index.js:18]:
```suggestion
  const project = await client.projects().create(organization_id, "project-name");
  console.log(project);
  const projects = await client.projects().list(organization_id);
```

File: languages/js/example/index.js [hunk 2]
```diff
@@ -11,22 +11,46 @@ async function main() {
...
+  const project_get = await client.projects().get(project.id);
+  console.log(project_get);
+
+  const updated_project = await client.projects().update(organization_id, project.id, "updated-project-name");
+  console.log(updated_project);
 
   const secret = await client
     .secrets()
-    .create("test-secret", "My secret!", "This is my secret", [project.id], organization_id);
+    .create(organization_id, "test-secret", "test-value", "test-note", [project.id]);
```
Comment (mzieniukbw) [inline languages/js/example/index.js:29]:
```suggestion
    .create(organization_id, "test-key", "test-value", "test-note", [project.id]);
```

File: languages/js/example/index.js [hunk 3]
```diff
@@ -11,22 +11,46 @@ async function main() {
...
   const secrets = await client.secrets().list(organization_id);
   console.log(secrets.data);
 
-  console.log(project, secret);
+  const secretData = await client.secrets().get(secret.id);
+  console.log(secrets.data);
```
Comment (mzieniukbw) [inline languages/js/example/index.js:34]:
```suggestion
  const secret_get = await client.secrets().get(secret.id);
  console.log(secret_get);
```

File: languages/js/example/index.js [hunk 4]
```diff
@@ -11,22 +11,46 @@ async function main() {
...
+
+  const updated_project = await client.projects().update(organization_id, project.id, "updated-project-name");
+  console.log(updated_project);
 
   const secret = await client
     .secrets()
-    .create("test-secret", "My secret!", "This is my secret", [project.id], organization_id);
+    .create(organization_id, "test-secret", "test-value", "test-note", [project.id]);
   const secrets = await client.secrets().list(organization_id);
   console.log(secrets.data);
```
Comment (mzieniukbw) [inline languages/js/example/index.js:31]: We `console.log` response one time and response's `.data` second time. Can we do one or another consistently ?

File: languages/js/example/index.js [hunk 5]
```diff
@@ -11,22 +11,46 @@ async function main() {
...
 
-  console.log(project, secret);
+  const secretData = await client.secrets().get(secret.id);
+  console.log(secrets.data);
+
+  const secretsByIds = await client.secrets().getByIds([secret.id]);
```
Comment (mzieniukbw) [inline languages/js/example/index.js:36]: I am confused are we using this style `secretsByIds` or this naming style `secrets_by_ids` across the file ? Seems like both right now

File: languages/js/example/index.js [hunk 6]
```diff
@@ -11,22 +11,46 @@ async function main() {
...
+
+  const updated_secret = await client.secrets().update(organization_id, secret.id, "updated-secret", "updated-value", "updated-note", [project.id]);
+  console.log(updated_secret);
+
+  const sync_has_changes_after_update = await client.secrets().sync(organization_id, now);
+  console.log(sync_has_changes_after_update.hasChanges);
+
+  for (const secret of secrets.data) {
+    await client.secrets().delete([secret.id]);
+  }
```
Comment (mzieniukbw) [inline languages/js/example/index.js:51]: We should only delete secret that we created, not all of them
```suggestion
  await client.secrets().delete([secret.id]);
```

File: languages/js/example/index.js [hunk 7]
```diff
@@ -11,22 +11,46 @@ async function main() {
...
+  console.log(secrets.data);
+
+  const secretsByIds = await client.secrets().getByIds([secret.id]);
+  console.log(secretsByIds.data);
+
+  const secret_sync_has_changes = await client.secrets().sync(organization_id, null);
+  const now = new Date();
+  console.log(secret_sync_has_changes.hasChanges);
+
+  const updated_secret = await client.secrets().update(organization_id, secret.id, "updated-secret", "updated-value", "updated-note", [project.id]);
```
Comment (mzieniukbw) [inline languages/js/example/index.js:43]:
```suggestion
  const updated_secret = await client.secrets().update(organization_id, secret.id, "updated-key", "updated-value", "updated-note", [project.id]);
```

**Review by mzieniukbw — CHANGES_REQUESTED** [review]
Example file needs some work on consistency

**Comment by coltonhurst** [comment]
@mzieniukbw thanks for the comments, I obviously requested a review too soon. The example has been updated in: https://github.com/bitwarden/sdk/pull/1044/commits/84bd3205d080e321c1f1ccabe826175d0823d8a5

**Review by mzieniukbw — APPROVED** [review]
LGTM!

File: languages/js/INSTALL.md [hunk 8]
```diff
@@ -0,0 +1,11 @@
+# WASM SDK Instructions
+
+## Steps
+
+1. Run the wasm build script: `./sdk/crates/bitwarden-wasm/build.sh`
```
Comment (Thomas-Avery) [inline languages/js/INSTALL.md:5]: Should we link to the wasm read me here? 

https://github.com/bitwarden/sdk/blob/main/crates/bitwarden-wasm/README.md

It requires installing some extra stuff to run that build script.

File: languages/js/sdk-client/src/client.ts [hunk 9]
```diff
@@ -131,6 +125,30 @@ export class SecretsClient {
 
     return handleResponse(Convert.toResponseForSecretsDeleteResponse(response));
   }
+
+  async sync(organizationId: string, lastSyncedDate: Date): Promise<SecretsSyncResponse> {
```
Comment (Thomas-Avery) [inline languages/js/sdk-client/src/client.ts:129]:
```suggestion
  async sync(organizationId: string, lastSyncedDate?: Date): Promise<SecretsSyncResponse> {
```

I believe passing the `lastSyncedDate` is optional.

**Review by Thomas-Avery — CHANGES_REQUESTED** [review]

File: languages/js/sdk-client/src/client.ts (hunk 9 above)
Comment (coltonhurst) [reply languages/js/sdk-client/src/client.ts]: Good call, resolved in https://github.com/bitwarden/sdk/pull/1044/commits/7f576ceabf2ecd736d2b57d8b8e2fabed46ee769

File: languages/js/INSTALL.md (hunk 8 above)
Comment (coltonhurst) [reply languages/js/INSTALL.md]: Resolved in: https://github.com/bitwarden/sdk/pull/1044/commits/7cb94c737f34a5f5fbb47420ff08ad2fd9802e0c

**Review by Thomas-Avery — DISMISSED** [review]
LGTM

**Review by dani-garcia — APPROVED** [review]

**Pull request closed without merging**